### PR TITLE
[Charmhub] Use refresh for resolving charm channels

### PR DESCRIPTION
**Pull request by SimonRichardson**
<strike>Requires #12556 to land, once done I'll rebase.</strike>

This introduces the concept of resolving charms with the preferred
channel via the refresh API. This removes the idea of using the info API
to get the possible charm or bundle revisions. Currently, we don't
support bundles as the API doesn't support bundles.

The code is actually quite simple and we can drop the charmhub selector
code that was required to source bundle download URLs.

The only difficulty now is ensuring that we provide all the right
updates to the charm URL and origin with the limited data that we have.

## Notes

 1. Bundles are not supported in this PR - the API isn't ready yet.
 2. Changes to the modeling now means the data in mongo reflects what you downloaded and not what the application series is.

## QA steps

Test every charm you can. Use the force as well.

```sh
$ juju add-model one --config charmhub-url="https://api.staging.charmhub.io"
$ juju deploy redis
$ juju deploy redis redis-focal --series=focal
ERROR series "focal" not supported by charm, supported series are: trusty. Use --force to deploy the charm anyway.
$ juju deploy redis redis-focal --series=focal --force
```

### CODE REVIEW

**Comment by hmlanigan** [comment]
QA
- `juju deploy ubuntu-juju-qa` as expected, deployed with bionic and stable, rev 14
- `juju deploy ubuntu-juju-qa --channel candidate` deployed xenial and candidate, rev 16 - expected focal due to default series and order of series in metadata.yaml
- `juju deploy ubuntu-juju-qa --channel edge` deployed xenial and edge, rev 17, expected groovy due to order of series in metadata.yaml, or at least focal, but not xenial.  Though with recent changes arguable expectation.
- `juju deploy ubuntu-juju-qa --channel edge --series groovy` as expected, deployed groovy and edge, rev 17

**Comment by hmlanigan** [comment]
Info on ubuntu-juju-qa charm used:
Revision Version Channel          Series in metadata.yaml
12       6       2.0/stable       disco, bionic, xenial, trusty
13       7       2.0/edge         bionic, xenial, trusty
14       8       latest/stable    bionic, xenial, trusty
16       9       latest/candidate focal, bionic, xenial, trusty
17      10       latest/edge     groovy, focal, bionic, xenial, trusty

**Comment by SimonRichardson** [comment]
@hmlanigan from your QA steps I believe this one's wrong

> juju deploy ubuntu-juju-qa as expected, deployed with bionic and stable, rev 14

This should be:
    juju deploy ubuntu-juju-qa deployed with focal and stable rev 2

**Comment by hmlanigan** [comment]
$ juju deploy ubuntu-juju-qa --channel 2.0/edge edge-not-latest
ERROR refresh retry: No revision was found in the Store.

$ juju deploy ubuntu-juju-qa --channel 2.0/edge edge-not-latest --series bionic
Located charm "ubuntu-juju-qa" in charm-hub, revision 13
Deploying "edge-not-latest" from charm-hub charm "ubuntu-juju-qa", revision 13 in channel 2.0/edge

**Comment by SimonRichardson** [comment]
@hmlanigan works as designed.

```sh
	
2:37 PM



 curl -XPOST -s https://api.staging.charmhub.io/v2/charms/refresh -H 'Content-type: application/json' -d '{"context": [], "actions": [{"name": "ubuntu-juju-qa", "platform": {"architecture": "amd64", "os": "NA", "series": "NA"}, "channel": "2.0/edge", "action": "install", "instance-key": "a-test"}]}' | jq

{
  "error-list": [],
  "results": [
    {
      "charm": null,
      "error": {
        "code": "invalid-charm-platform",
        "extra": {
          "default-platforms": [
            {
              "architecture": "amd64",
              "os": "ubuntu",
              "series": "focal"
            }
          ]
        },
        "message": "Instance key 'a-test' invalid 'platform' in 'install' for charm_id=YwXwMO6R3YyWflI61y2YFppHhGr6RO66 name=ubuntu-juju-qa"
      },
      "id": "YwXwMO6R3YyWflI61y2YFppHhGr6RO66",
      "instance-key": "a-test",
      "name": "ubuntu-juju-qa",
      "released-at": null,
      "result": "error"
    }
  ]
}
```

File: apiserver/facades/client/charms/charmhubrepo_test.go [hunk 1]
```diff
@@ -4,180 +4,506 @@
...
-	cp := channelPlatform{
-		Platform: corecharm.MustParsePlatform("amd64"),
+func (refreshConfigSuite) TestRefreshByRevision(c *gc.C) {
+	revision := 1
+	curl := charm.MustParseURL("ch:wordpress")
+	platform := corecharm.MustParsePlatform("amd64/ubuntu/focal")
+	channel := corecharm.MustParseChannel("latest/stable").Normalize()
+	origin := corecharm.Origin{
+		Platform: platform,
+		Channel:  &channel,
```
Comment (hmlanigan) [inline apiserver/facades/client/charms/charmhubrepo_test.go:278]: test has inacurate expectations, calls to refresh will fail if both channel and revision are provided:
 "error": {
        "code": "revision-conflict",
        "message": "Both revision and channel are present in the request for this Charm. It should be one or the other exclusively."
      },

File: apiserver/facades/client/charms/charmhubrepo_test.go [hunk 2]
```diff
@@ -4,180 +4,506 @@
...
+	revision := 1
+	curl := charm.MustParseURL("ch:wordpress")
+	platform := corecharm.MustParsePlatform("amd64/ubuntu/focal")
+	channel := corecharm.MustParseChannel("latest/stable").Normalize()
+	origin := corecharm.Origin{
+		ID:       id,
+		Platform: platform,
+		Channel:  &channel,
+		Revision: &revision,
 	}
```
Comment (hmlanigan) [inline apiserver/facades/client/charms/charmhubrepo_test.go:316]: test has inacurate expectations, calls to refresh will fail if both channel and revision are provided: "error": { "code": "revision-conflict", "message": "Both revision and channel are present in the request for this Charm. It should be one or the other exclusively." },

**Review by hmlanigan — APPROVED** [review]
A few changes please.  QA went well, though ran into the api response bug: https://bugs.launchpad.net/snapstore-server/+bug/1913789

File: apiserver/facades/client/charms/charmhubrepo.go [hunk 3]
```diff
@@ -22,88 +23,121 @@ import (
...
 	}
 
-	platform, err := makePlatform(origin)
-	if err != nil {
-		return nil, params.CharmOrigin{}, nil, errors.Trace(err)
-	}
+	if resErr := refreshRes.Error; resErr != nil {
+		switch resErr.Code {
+		case transport.ErrorCodeInvalidCharmPlatform:
+			logger.Errorf("Invalid charm platform %q %v - Default Platforms: %v", curl, origin, resErr.Extra.DefaultPlatforms)
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply apiserver/facades/client/charms/charmhubrepo.go:71]: I'll drop this down to a trace, it's really useful for when debugging.

File: apiserver/facades/client/charms/charmhubrepo.go [hunk 4]
```diff
@@ -22,88 +23,121 @@ import (
...
+	revision := refreshRes.Entity.Revision
+	resCurl := curl.
+		WithSeries(origin.Series).
+		WithArchitecture(origin.Architecture).
+		WithRevision(revision)
+
+	// TODO (stickupkid): This is currently hardcoded as the API doesn't support
+	// bundles.
+	resOrigin := origin
+	resOrigin.Type = "charm"
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply apiserver/facades/client/charms/charmhubrepo.go:128]: I'll fix this in the bundle refresh PR.

File: apiserver/facades/client/charms/charmhubrepo.go [hunk 5]
```diff
@@ -22,88 +23,121 @@ import (
...
+	resOrigin.ID = refreshRes.ID
+	resOrigin.Hash = refreshRes.Entity.Download.HashSHA256
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply apiserver/facades/client/charms/charmhubrepo.go:130]: Agreed :100:

**Comment by SimonRichardson** [comment]
$$merge$$

**Comment by SimonRichardson** [comment]
$$merge$$

> FAIL: flag_test.go:94: WorkerSuite.TestErrRefresh

**Pull request merged**